### PR TITLE
[SPARK-31179][Core] Fast fail the connection while last connection failed in fast fail time window

### PR DESCRIPTION
**Pull request by Sandykolas**
### What changes were proposed in this pull request?

For TransportFactory, the requests sent to the same address share a clientPool.
    Specially, when the io.numConnectionPerPeer is 1, these requests would share a same client.
    When this address is unreachable, the createClient operation would be still timeout.
    And these requests would block each other during createClient, because there is a lock for this shared client.
    It would cost connectionNum \* connectionTimeOut \* maxRetry to retry, and then fail the task.

### Why are the changes needed?
   I set a fastFail time window for the clientPool, if the last connection failed in this time window, the new connection would fast fail.

### Does this PR introduce any user-facing change?
dev only

### How was this patch tested?
 No.
 Existing UT.

### CODE REVIEW

File: common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java [hunk 1]
```diff
@@ -184,11 +186,30 @@ public TransportClient createClient(String remoteHost, int remotePort)
           logger.info("Found inactive connection to {}, creating a new one.", resolvedAddress);
         }
       }
-      clientPool.clients[clientIndex] = createClient(resolvedAddress);
+      if (fastFail && System.currentTimeMillis() - clientPool.lastConnectionFailed
+              < fastFailTimeWindow) {
```
Comment (yaooqinn) [inline common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java:190]: indent

File: common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java [hunk 2]
```diff
@@ -184,11 +186,30 @@ public TransportClient createClient(String remoteHost, int remotePort)
...
     }
   }
 
+
+  public TransportClient createClient(String remoteHost, int remotePort)
+    throws IOException, InterruptedException {
+      return createClient(remoteHost, remotePort, false);
+  }
+
+
```
Comment (yaooqinn) [inline common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java:212]: blank line

Comment (yaooqinn) [inline common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java:212]: https://github.com/apache/spark/pull/27943/files#diff-42200a79a017015382cc1819ece43f77c30f40f357475d7979133b0906789ccaR245-R248 do you really follow the original PR?

File: common/network-common/src/test/java/org/apache/spark/network/TransportClientFactorySuite.java [hunk 3]
```diff
@@ -44,6 +41,7 @@
 import org.apache.spark.network.util.MapConfigProvider;
 import org.apache.spark.network.util.JavaUtils;
 import org.apache.spark.network.util.TransportConf;
+import org.junit.rules.ExpectedException;
```
Comment (yaooqinn) [inline common/network-common/src/test/java/org/apache/spark/network/TransportClientFactorySuite.java:44]: style: this belongs to the third-party group

File: common/network-common/src/test/java/org/apache/spark/network/TransportClientFactorySuite.java [hunk 4]
```diff
@@ -26,10 +26,7 @@
 import java.util.Set;
 import java.util.concurrent.atomic.AtomicInteger;
 
-import org.junit.After;
-import org.junit.Assert;
-import org.junit.Before;
-import org.junit.Test;
+import org.junit.*;
```
Comment (yaooqinn) [inline common/network-common/src/test/java/org/apache/spark/network/TransportClientFactorySuite.java:29]: unnecessary change

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleClient.java [hunk 5]
```diff
@@ -95,14 +95,14 @@ public void fetchBlocks(
...
     try {
+      int maxRetries = conf.maxIORetries();
       RetryingBlockFetcher.BlockFetchStarter blockFetchStarter =
           (blockIds1, listener1) -> {
-            TransportClient client = clientFactory.createClient(host, port);
-            new OneForOneBlockFetcher(client, appId, execId,
-              blockIds1, listener1, conf, downloadFileManager).start();
+         TransportClient client = clientFactory.createClient(host, port, maxRetries > 0);
+         new OneForOneBlockFetcher(client, appId, execId,
+                  blockIds1, listener1, conf, downloadFileManager).start();
```
Comment (yaooqinn) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleClient.java:103]: indent

Comment (yaooqinn) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleClient.java:103]: indent

File: core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala [hunk 6]
```diff
@@ -109,17 +109,18 @@ private[spark] class NettyBlockTransferService(
...
+          val client = clientFactory.createClient(host, port, maxRetries > 0)
           new OneForOneBlockFetcher(client, appId, execId, blockIds, listener,
             transportConf, tempFileManager).start()
         }
       }
 
-      val maxRetries = transportConf.maxIORetries()
       if (maxRetries > 0) {
         // Note this Fetcher will correctly handle maxRetries == 0; we avoid it just in case there's
+
```
Comment (yaooqinn) [inline core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala:123]: unnecessary

File: core/src/test/scala/org/apache/spark/storage/BlockManagerReplicationSuite.scala [hunk 7]
```diff
@@ -421,65 +421,65 @@ class BlockManagerReplicationSuite extends BlockManagerReplicationBehavior {
...
-
-    // Make sure all locks have been released.
-    eventually(timeout(1000 milliseconds), interval(10 milliseconds)) {
-      initialStores.filter(bm => newLocations.contains(bm.blockManagerId)).foreach { bm =>
-        assert(bm.blockInfoManager.getTaskLockCount(BlockInfo.NON_TASK_WRITER) === 0)
-      }
-    }
-  }
-}
+//  class BlockManagerProactiveReplicationSuite extends BlockManagerReplicationBehavior {
```
Comment (yaooqinn) [inline core/src/test/scala/org/apache/spark/storage/BlockManagerReplicationSuite.scala:424]: bring this test back？

**Review by Sandykolas — COMMENTED** [review]
For TransportFactory, the client sent to the same address share a clientPool.
    Specially, when the io.numConnectionPerPeer is 1, these requests would share a same client.
    When this address is unreachable, the createClient operation would be still timeout.
    And these requests would block each other during createClient, because there is a lock for this shared client.
    It would cost connectionNum \* connectionTimeOut \* maxRetry to retry, and then fail the task.

    It fact, it is expected that this task could fail in connectionTimeOut * maxRetry.

    In this PR, I set a fastFail time window for the clientPool, if the last connection failed in this time window, the new connection would fast fail.

    It can save time for some cases.

**Comment by Sandykolas** [comment]
For TransportFactory, the requests sent to the same address share a clientPool.
    Specially, when the io.numConnectionPerPeer is 1, these requests would share a same client.
    When this address is unreachable, the createClient operation would be still timeout.
    And these requests would block each other during createClient, because there is a lock for this shared client.
    It would cost connectionNum \* connectionTimeOut \* maxRetry to retry, and then fail the task.
    It fact, it is expected that this task could fail in connectionTimeOut * maxRetry.
    In this PR, I set a fastFail time window for the clientPool, if the last connection failed in this time window, the new connection would fast fail.
    It can save time for some cases.

File: common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java (hunk 1 above)
Comment (Sandykolas) [reply common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java:190]: we have set an fastFailTimeWindow  to limits the fast links linked times,

Comment (yaooqinn) [reply common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java:190]: 我只是说这里代码缩进可能有问题而已。请fix。。。

Comment (Sandykolas) [reply common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java:190]: I have resloved the scalastyle problem, so sorry, wo can't understand your meaning

File: common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java (hunk 2 above)
Comment (Sandykolas) [reply common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java:212]: TransportClinetFactory to create createClient in order to linked remotehost.

Comment (Sandykolas) [reply common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java:212]: yeah ,I have follows the original PR，the orihinal PR added an paramer to  acquired  the results that links the remoteHost.

Comment (yaooqinn) [reply common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java:212]: https://github.com/apache/spark/pull/27943/files#diff-42200a79a017015382cc1819ece43f77c30f40f357475d7979133b0906789ccaR246-R247

the original PR have these changes

File: common/network-common/src/test/java/org/apache/spark/network/TransportClientFactorySuite.java (hunk 3 above)
Comment (Sandykolas) [reply common/network-common/src/test/java/org/apache/spark/network/TransportClientFactorySuite.java:44]: sorrry, i have resolved the  problem about scalastyle

File: common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java (hunk 2 above)
Comment (Sandykolas) [reply common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java:212]: At spark-2.3.2 nerver changes ,In our daily working, we  have find the issue that when the io.numConnectionPerPeer is 1, these requests would share a same client.When this address is unreachable, the createClient operation would be still timeout.
in order to sovled the problem that we have to set an fastFailTimeWindow to  limits that the job running times out ,
so we have set the fastfail to acquie the results ( the fastfail being boolean),
so we can acquire the results that clientpool creates the remotehost

File: common/network-common/src/test/java/org/apache/spark/network/TransportClientFactorySuite.java (hunk 4 above)
Comment (Sandykolas) [reply common/network-common/src/test/java/org/apache/spark/network/TransportClientFactorySuite.java:29]: yes , I have solved the problem

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleClient.java (hunk 5 above)
Comment (Sandykolas) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleClient.java:103]: the scalastyle problem have resolved

File: core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala (hunk 6 above)
Comment (Sandykolas) [reply core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala:123]: yeah, I will resoved the problem that scalastyle

File: core/src/test/scala/org/apache/spark/storage/BlockManagerReplicationSuite.scala (hunk 7 above)
Comment (Sandykolas) [reply core/src/test/scala/org/apache/spark/storage/BlockManagerReplicationSuite.scala:424]: yeah ,I will recover the test demo

**Comment by Sandykolas** [comment]
What changes were proposed in this pull request?
For TransportFactory, the requests sent to the same address share a clientPool.
Specially, when the io.numConnectionPerPeer is 1, these requests would share a same client.
When this address is unreachable, the createClient operation would be still timeout.
And these requests would block each other during createClient, because there is a lock for this shared client.
It would cost connectionNum * connectionTimeOut * maxRetry to retry, and then fail the task.

Why are the changes needed?
I set a fastFail time window for the clientPool, if the last connection failed in this time window, the new connection would fast fail.

Does this PR introduce any user-facing change?
dev only

How was this patch tested?
No.
Existing UT.

**Comment by Sandykolas** [comment]
What changes were proposed in this pull request?
For TransportFactory, the requests sent to the same address share a clientPool.
Specially, when the io.numConnectionPerPeer is 1, these requests would share a same client.
When this address is unreachable, the createClient operation would be still timeout.
And these requests would block each other during createClient, because there is a lock for this shared client.
It would cost connectionNum * connectionTimeOut * maxRetry to retry, and then fail the task.

Why are the changes needed?
I set a fastFail time window for the clientPool, if the last connection failed in this time window, the new connection would fast fail.

Does this PR introduce any user-facing change?
dev only

How was this patch tested?
No.
Existing UT.

File: common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java (hunk 1 above)
Comment (Sandykolas) [reply common/network-common/src/main/java/org/apache/spark/network/client/TransportClientFactory.java:190]: this is the result of a new line that not cause scalastyle to report an error，
but I will set it in tab format

File: common/network-common/src/test/java/org/apache/spark/network/TransportClientFactorySuite.java (hunk 4 above)
Comment (Sandykolas) [reply common/network-common/src/test/java/org/apache/spark/network/TransportClientFactorySuite.java:29]: followed the company code and change it

**Review by Sandykolas — COMMENTED** [review]
remove tab space cause error

**Review by Sandykolas — COMMENTED** [review]
remove tab space cause error
